In [3]:
import yfinance as yf
import pandas as pd

# Define the ticker symbols for NASDAQ and NSE indices
nasdaq_symbol = "^IXIC"  # NASDAQ ticker symbol
nse_symbol = "^NSEI"  # NSE ticker symbol

# Set the start and end dates for the historical data
start_date = "2010-01-01"  # Replace with your desired start date
end_date = "2023-05-01"  # Replace with your desired end date

# Use yf.download() to retrieve the historical data
nasdaq_data = yf.download(nasdaq_symbol, start=start_date, end=end_date)
nse_data = yf.download(nse_symbol, start=start_date, end=end_date)

# Convert the data to pandas DataFrames
nasdaq_df = pd.DataFrame(nasdaq_data)
nse_df = pd.DataFrame(nse_data)

# Display the first few rows of each DataFrame
print("NASDAQ data:")
print(nasdaq_df.head())
print("\nNSE data:")
print(nse_df.head())

# Save the data as CSV files
nasdaq_df.to_csv("nasdaq_data.csv", index=False)
nse_df.to_csv("nse_data.csv", index=False)
import pandas as pd

nasdaq_df = pd.read_csv("nasdaq_data.csv")
nse_df = pd.read_csv("nse_data.csv")
nasdaq_close = nasdaq_df["Close"]
nse_close = nse_df["Close"]
correlation_coefficient = nasdaq_close.corr(nse_close)



[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
NASDAQ data:
                   Open         High          Low        Close    Adj Close  \
Date                                                                          
2010-01-04  2294.409912  2311.149902  2294.409912  2308.419922  2308.419922   
2010-01-05  2307.270020  2313.729980  2295.620117  2308.709961  2308.709961   
2010-01-06  2307.709961  2314.070068  2295.679932  2301.090088  2301.090088   
2010-01-07  2298.090088  2301.300049  2285.219971  2300.050049  2300.050049   
2010-01-08  2292.239990  2317.600098  2290.610107  2317.169922  2317.169922   

                Volume  
Date                    
2010-01-04  1931380000  
2010-01-05  2367860000  
2010-01-06  2253340000  
2010-01-07  2270050000  
2010-01-08  2145390000  

NSE data:
                   Open         High          Low        Close    Adj Close  \
Date                          

In [4]:
print("Correlation coefficient:", correlation_coefficient)


Correlation coefficient: 0.9607951810049128


The correlation coefficient of 0.9608 is very close to 1, which indicates a high degree of correlation. This means that the movements of the NASDAQ and NSE indices are highly synchronized, and they tend to move in the same direction with a similar magnitude.

Overall, the analysis of the correlation coefficient suggests a strong positive relationship between the NASDAQ and NSE indices, indicating that they are closely related and move in a similar pattern.

however ,correlation does not imply causation. While there is a strong correlation between the two indices, it does not necessarily mean that one index causes the other to move. It is important to consider other factors and conduct further analysis to understand the underlying dynamics and potential causality.


In [7]:
nasdaq_df['NASDAQ_PctChange'] = nasdaq_df['Close'].pct_change()
nse_df['NSE_PctChange'] = nse_df['Close'].pct_change()
nse_df['Lagged_NASDAQ_PctChange'] = nasdaq_df['NASDAQ_PctChange'].shift(1)
nse_df['LeadLag_Difference'] = nse_df['Lagged_NASDAQ_PctChange'] - nse_df['NSE_PctChange']


In [24]:
import pandas as pd

# Print the column names
print(nse_df.columns)


Index(['Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume', 'NSE_PctChange',
       'Lagged_NASDAQ_PctChange', 'LeadLag_Difference'],
      dtype='object')


In [11]:
import pandas as pd

# Print the data types of all columns
print(nse_df.dtypes)


Open                       float64
High                       float64
Low                        float64
Close                      float64
Adj Close                  float64
Volume                       int64
NSE_PctChange              float64
Lagged_NASDAQ_PctChange    float64
LeadLag_Difference         float64
dtype: object


In [21]:

import talib


# Define the symbol or ticker for the desired stock
symbol = 'AAPL'

# Define the start and end dates for the historical data
start_date = '2010-01-01'
end_date = '2023-05-01'

# Fetch the historical price data using yfinance
df = yf.download(symbol, start=start_date, end=end_date)

# Print the first few rows of the DataFrame to verify the data
print(df.head())


# Load the historical price data into a Pandas DataFrame
# Replace 'df' with the actual DataFrame containing the price data
df = pd.read_csv('historical_prices.csv')

period = 20  # Define the period for the Keltner Channel
atr_period = 10  # Define the period for calculating Average True Range (ATR)
bb_period = 20  # Define the period for the Bollinger Bands
std_dev = 2  # Define the number of standard deviations
macd_fast_period = 12  # Define the period for the fast line of MACD
macd_slow_period = 26  # Define the period for the slow line of MACD
macd_signal_period = 9  # Define the period for the signal line of MACD

# Calculate the necessary components for the Keltner Channel
high = df['High']
low = df['Low']
close = df['Close']
atr = talib.ATR(high, low, close, timeperiod=atr_period)  # Calculate ATR
keltner_upper = talib.SMA(close, timeperiod=period) + atr * 1.5
keltner_middle = talib.SMA(close, timeperiod=period)
keltner_lower = talib.SMA(close, timeperiod=period) - atr * 1.5

# Calculate the Bollinger Bands
bb_upper, bb_middle, bb_lower = talib.BBANDS(close, timeperiod=bb_period, nbdevup=std_dev, nbdevdn=std_dev)

# Calculate the MACD
macd, macd_signal, macd_hist = talib.MACD(close, fastperiod=macd_fast_period, slowperiod=macd_slow_period,
                                         signalperiod=macd_signal_period)

# Add the Keltner Channel, Bollinger Bands, and MACD values to the DataFrame
df['Keltner_Upper'] = keltner_upper
df['Keltner_Middle'] = keltner_middle
df['Keltner_Lower'] = keltner_lower
df['BB_Upper'] = bb_upper
df['BB_Middle'] = bb_middle
df['BB_Lower'] = bb_lower
df['MACD'] = macd
df['MACD_Signal'] = macd_signal
df['MACD_Histogram'] = macd_hist

# Analyze the indicators and their values for further analysis or trading strategy development


[*********************100%***********************]  1 of 1 completed
                  Open        High         Low       Close   Adj Close  \
Date                                                                     
2022-01-03  177.830002  182.880005  177.710007  182.009995  180.434296   
2022-01-04  182.630005  182.940002  179.119995  179.699997  178.144287   
2022-01-05  179.610001  180.169998  174.639999  174.919998  173.405670   
2022-01-06  172.699997  175.300003  171.639999  172.000000  170.510941   
2022-01-07  172.889999  174.139999  171.029999  172.169998  170.679489   

               Volume  
Date                   
2022-01-03  104487900  
2022-01-04   99310400  
2022-01-05   94537600  
2022-01-06   96904000  
2022-01-07   86709100  


FileNotFoundError: [Errno 2] No such file or directory: 'historical_prices.csv'

In [18]:
import os
print(os.getcwd())


C:\Users\hp\Downloads


In [22]:
import yfinance as yf
import talib

# Define the symbol or ticker for the index
symbol = '^GSPC'

# Define the start and end dates for the historical data
start_date = '2018-01-01'
end_date = '2022-12-31'

# Fetch the historical price data using yfinance
data = yf.download(symbol, start=start_date, end=end_date)

# Define the parameter space for the period of the Keltner Channel
periods = [10, 20, 30, 40, 50]

# Define the optimization function
def optimize_parameters(period):
    # Calculate the Keltner Channel with the given period
    close = data['Close']
    atr = talib.ATR(data['High'], data['Low'], close, timeperiod=10)
    keltner_upper = talib.SMA(close, timeperiod=period) + atr * 1.5
    keltner_middle = talib.SMA(close, timeperiod=period)
    keltner_lower = talib.SMA(close, timeperiod=period) - atr * 1.5

    # Define your strategy and evaluate its performance
    # Use the metrics coded in your last assignment to calculate the performance metric

    # Return the performance metric
    return performance_metric

best_performance = float('-inf')
best_period = None

# Perform parameter optimization
for period in periods:
    performance = optimize_parameters(period)
    if performance > best_performance:
        best_performance = performance
        best_period = period

# Document the optimized parameters
print("Optimized parameters:")
print("Period for Keltner Channel:", best_period)
print("Performance metric:", best_performance)


[*********************100%***********************]  1 of 1 completed


NameError: name 'performance_metric' is not defined